README.md
# oauth2orize-openid

[OAuth2orize](https://github.com/passport-next/oauth2orize) extensions providing
support for [OpenID Connect](https://openid.net/connect/).

## Examples

This [example](https://github.com/gerges-beshay/oauth2orize-openid-examples) demonstrates
basic usage of the api.

CHANGELOG.md
# 2.0.0 (2018-11-20)

* Added CHANGELOG.md @adamhathcock
* Updated deps, replacing utils-merge with lodash @adamhathcock
* Added explicit dependency on passport-next/oauth2orize for AuthorizationError type
* Updated README.md package.json etc for migration to passport-next @adamhathcock


lib/grant/codeToken.js
/**
 * Module dependencies.
 */
var _ = require('lodash')
  , AuthorizationError = require('../errors/authorizationerror');


/**
 * Handles requests to obtain a response with an access token and authorization
 * code.
 *
 * References:
 *  - [OpenID Connect Standard 1.0 - draft 21](http://openid.net/specs/openid-connect-standard-1_0.html)
 *  - [OpenID Connect Messages 1.0 - draft 20](http://openid.net/specs/openid-connect-messages-1_0.html)
 *  - [OAuth 2.0 Multiple Response Type Encoding Practices - draft 08](http://openid.net/specs/oauth-v2-multiple-response-types-1_0.html)
 *
 * @param {Object} options
 * @param {Function} issue
 * @return {Object} module
 * @api public
 */
module.exports = function(options, issueToken, issueCode) {
  if (typeof options == 'function') {
    issueCode = issueToken;
    issueToken = options;
    options = undefined;
  }
  options = options || {};
  
  if (!issueToken) throw new TypeError('oauth2orize-openid.codeToken grant requires an issueToken callback');
  if (!issueCode) throw new TypeError('oauth2orize-openid.codeToken grant requires an issueCode callback');

  var modes = options.modes || {};
  if (!modes.fragment) {
    modes.fragment = require('../response/fragment');
  }

  // For maximum flexibility, multiple scope spearators can optionally be
  // allowed.  This allows the server to accept clients that separate scope
  // with either space or comma (' ', ',').  This violates the specification,
  // but achieves compatibility with existing client libraries that are already
  // deployed.
  var separators = options.scopeSeparator || ' ';
  if (!Array.isArray(separators)) {
    separators = [ separators ];
  }
  
  
  /* Parse requests that request `code token` as `response_type`.
   *
   * @param {http.ServerRequest} req
   * @api public
   */
  function request(req) {
    var clientID = req.query['client_id']
      , redirectURI = req.query['redirect_uri']
      , scope = req.query['scope']
      , state = req.query['state'];
      
    if (!clientID) { throw new AuthorizationError('Missing required parameter: client_id', 'invalid_request'); }
    if (typeof clientID !== 'string') { throw new AuthorizationError('Invalid parameter: client_id must be a string', 'invalid_request'); }
    
    if (scope) {
      if (typeof scope !== 'string') {
        throw new AuthorizationError('Invalid parameter: scope must be a string', 'invalid_request');
      }
      
      for (var i = 0, len = separators.length; i < len; i++) {
        var separated = scope.split(separators[i]);
        // only separate on the first matching separator.  this allows for a sort
        // of separator "priority" (ie, favor spaces then fallback to commas)
        if (separated.length > 1) {
          scope = separated;
          break;
        }
      }
      
      if (!Array.isArray(scope)) { scope = [ scope ]; }
    }
    
    return {
      clientID: clientID,
      redirectURI: redirectURI,
      scope: scope,
      state: state
    }
  }
  
  /* Sends responses to transactions that request `code token` as `response_type`.
   *
   * @param {Object} txn
   * @param {http.ServerResponse} res
   * @param {Function} complete
   * @param {Function} next
   * @api public
   */
  function response(txn, res, complete, next) {
    if (next === undefined) {
      next = complete;
      complete = function(cb) { return cb(); };
    }

    var mode = 'fragment'
      , respond;
    if (txn.req && txn.req.responseMode) {
      mode = txn.req.responseMode;
    }
    respond = modes[mode];
    
    if (!respond) {
      // http://lists.openid.net/pipermail/openid-specs-ab/Week-of-Mon-20140317/004680.html
      return next(new AuthorizationError('Unsupported response mode: ' + mode, 'unsupported_response_mode', null, 501));
    }
    if (respond && respond.validate) {
      try {
        respond.validate(txn);
      } catch(ex) {
        return next(ex);
      }
    }

    if (!txn.res.allow) {
      var params = { error: 'access_denied' };
      if (txn.req && txn.req.state) { params.state = txn.req.state; }
      return respond(txn, res, params);
    }
    
    function doIssueCode(tok) {
      function issued(err, code) {
        if (err) { return next(err); }
        if (!code) { return next(new AuthorizationError('Request denied by authorization server', 'access_denied')); }
      
        tok['code'] = code;

        complete(function(err) {
          if (err) { return next(err); }
          return respond(txn, res, tok);
        });
      }
    
      try {
        // NOTE: To facilitate code reuse, the `issueCode` callback should
        //       interoperate with the `issue` callback implemented by
        //       `oauth2orize.grant.code`.
        
        var arity = issueCode.length;
        if (arity == 7) {
          var locals = txn.locals || {};
          locals.accessToken = tok.access_token;
          issueCode(txn.client, txn.req.redirectURI, txn.user, txn.res, txn.req, locals, issued);
        } else if (arity == 6) {
          issueCode(txn.client, txn.req.redirectURI, txn.user, txn.res, txn.req, issued);
        } else if (arity == 5) {
          issueCode(txn.client, txn.req.redirectURI, txn.user, txn.res, issued);
        } else { // arity == 4
          issueCode(txn.client, txn.req.redirectURI, txn.user, issued);
        }
      } catch (ex) {
        return next(ex);
      }
    }
    
    function doIssueToken() {
      function issued(err, accessToken, params) {
        if (err) { return next(err); }
        if (!accessToken) { return next(new AuthorizationError('Request denied by authorization server', 'access_denied')); }
      
        var tok = {};
        tok['access_token'] = accessToken;
        if (params) { _.merge(tok, params); }
        tok['token_type'] = tok['token_type'] || 'Bearer';
        if (txn.req && txn.req.state) { tok['state'] = txn.req.state; }
      
        doIssueCode(tok);
      }
      
      try {
        // NOTE: To facilitate code reuse, the `issueToken` callback should
        //       interoperate with the `issue` callback implemented by
        //       `oauth2orize.grant.token`.
        
        var arity = issueToken.length;
        if (arity == 6) {
          issueToken(txn.client, txn.user, txn.res, txn.req, txn.locals, issued);
        } else if (arity == 5) {
          issueToken(txn.client, txn.user, txn.res, txn.req, issued);
        } else if (arity == 4) {
          issueToken(txn.client, txn.user, txn.res, issued);
        } else { // arity == 3
          issueToken(txn.client, txn.user, issued);
        }
      } catch (ex) {
        return next(ex);
      }
    }
    
    doIssueToken();
  }

  function errorHandler(err, txn, res, next) {
    var mode = 'fragment'
    , params = {}
    , respond;
    if (txn.req && txn.req.responseMode) {
      mode = txn.req.responseMode;
    }
    respond = modes[mode];

    if (!respond) {
      return next(err);
    }
    if (respond && respond.validate) {
      try {
        respond.validate(txn);
      } catch(ex) {
        return next(err);
      }
    }

    params.error = err.code || 'server_error';
    if (err.message) { params.error_description = err.message; }
    if (err.uri) { params.error_uri = err.uri; }
    if (txn.req && txn.req.state) { params.state = txn.req.state; }
    return respond(txn, res, params);
  }

  /**
   * Return `code token` grant module.
   */
  var mod = {};
  mod.name = 'code token';
  mod.request = request;
  mod.response = response;
  mod.error = errorHandler;
  return mod;
}


lib/grant/idTokenToken.js
/**
 * Module dependencies.
 */
var _ = require('lodash')
  , AuthorizationError = require('../errors/authorizationerror');


/**
 * Handles requests to obtain a response with an access token and ID token.
 *
 * References:
 *  - [OpenID Connect Standard 1.0 - draft 21](http://openid.net/specs/openid-connect-standard-1_0.html)
 *  - [OpenID Connect Messages 1.0 - draft 20](http://openid.net/specs/openid-connect-messages-1_0.html)
 *  - [OAuth 2.0 Multiple Response Type Encoding Practices - draft 08](http://openid.net/specs/oauth-v2-multiple-response-types-1_0.html)
 *
 * @param {Object} options
 * @param {Function} issue
 * @return {Object} module
 * @api public
 */
module.exports = function(options, issueToken, issueIDToken) {
  if (typeof options == 'function') {
    issueIDToken = issueToken;
    issueToken = options;
    options = undefined;
  }
  options = options || {};
  
  if (!issueToken) throw new TypeError('oauth2orize-openid.idTokenToken grant requires an issueToken callback');
  if (!issueIDToken) throw new TypeError('oauth2orize-openid.idTokenToken grant requires an issueIDToken callback');

  var modes = options.modes || {};
  if (!modes.fragment) {
    modes.fragment = require('../response/fragment');
  }

  // For maximum flexibility, multiple scope spearators can optionally be
  // allowed.  This allows the server to accept clients that separate scope
  // with either space or comma (' ', ',').  This violates the specification,
  // but achieves compatibility with existing client libraries that are already
  // deployed.
  var separators = options.scopeSeparator || ' ';
  if (!Array.isArray(separators)) {
    separators = [ separators ];
  }
  
  
  /* Parse requests that request `id_token token` as `response_type`.
   *
   * @param {http.ServerRequest} req
   * @api public
   */
  function request(req) {
    var clientID = req.query['client_id']
      , redirectURI = req.query['redirect_uri']
      , scope = req.query['scope']
      , state = req.query['state']
      , nonce = req.query['nonce'];
    
    if (!clientID) { throw new AuthorizationError('Missing required parameter: client_id', 'invalid_request'); }
    if (typeof clientID !== 'string') { throw new AuthorizationError('Invalid parameter: client_id must be a string', 'invalid_request'); }

    if (!nonce) { throw new AuthorizationError('Missing required parameter: nonce', 'invalid_request'); }
    if (typeof nonce !== 'string') { throw new AuthorizationError('Invalid parameter: nonce must be a string', 'invalid_request'); }
    
    if (scope) {
      if (typeof scope !== 'string') {
        throw new AuthorizationError('Invalid parameter: scope must be a string', 'invalid_request');
      }
      
      for (var i = 0, len = separators.length; i < len; i++) {
        var separated = scope.split(separators[i]);
        // only separate on the first matching separator.  this allows for a sort
        // of separator "priority" (ie, favor spaces then fallback to commas)
        if (separated.length > 1) {
          scope = separated;
          break;
        }
      }
      
      if (!Array.isArray(scope)) { scope = [ scope ]; }
    }
    
    return {
      clientID: clientID,
      redirectURI: redirectURI,
      scope: scope,
      state: state,
      nonce: nonce
    }
  }
  
  /* Sends responses to transactions that request `id_token token` as `response_type`.
   *
   * @param {Object} txn
   * @param {http.ServerResponse} res
   * @param {Function} complete
   * @param {Function} next
   * @api public
   */
  function response(txn, res, complete, next) {
    if (next === undefined) {
      next = complete;
      complete = function(cb) { return cb(); };
    }

    var mode = 'fragment'
      , respond;

    if (txn.req && txn.req.responseMode) {
      mode = txn.req.responseMode;
    }
    respond = modes[mode];
    
    if (!respond) {
      // http://lists.openid.net/pipermail/openid-specs-ab/Week-of-Mon-20140317/004680.html
      return next(new AuthorizationError('Unsupported response mode: ' + mode, 'unsupported_response_mode', null, 501));
    }
    if (respond && respond.validate) {
      try {
        respond.validate(txn);
      } catch(ex) {
        return next(ex);
      }
    }

    if (!txn.res.allow) {
      var params = { error: 'access_denied' };
      if (txn.req && txn.req.state) { params.state = txn.req.state; }
      return respond(txn, res, params);
    }
    
    function doIssueIDToken(tok) {
      function issued(err, idToken) {
        if (err) { return next(err); }
        if (!idToken) { return next(new AuthorizationError('Request denied by authorization server', 'access_denied')); }
      
        tok['id_token'] = idToken;
        if (txn.req && txn.req.state) { tok['state'] = txn.req.state; }

        complete(function(err) {
          if (err) { return next(err); }
          return respond(txn, res, tok);
        });
      }
    
      try {
        // NOTE: To facilitate code reuse, the `issueIDToken` callback should
        //       interoperate with the `issue` callback implemented by
        //       `oauth2orize-openid.grant.idToken`.
        
        var arity = issueIDToken.length;
        if (arity == 7) {
          issueIDToken(txn.client, txn.user, txn.res, txn.req, { accessToken: tok.access_token }, txn.locals, issued);
        } else if (arity == 6) {
          issueIDToken(txn.client, txn.user, txn.res, txn.req, { accessToken: tok.access_token }, issued);
        } else if (arity == 5) {
          issueIDToken(txn.client, txn.user, txn.res, txn.req, issued);
        } else { // arity == 4
          issueIDToken(txn.client, txn.user, txn.req, issued);
        }
      } catch (ex) {
        return next(ex);
      }
    }
    
    function doIssueToken() {
      function issued(err, accessToken, params) {
        if (err) { return next(err); }
        if (!accessToken) { return next(new AuthorizationError('Request denied by authorization server', 'access_denied')); }
      
        var tok = {};
        tok['access_token'] = accessToken;
        if (params) { _.merge(tok, params); }
        tok['token_type'] = tok['token_type'] || 'Bearer';
        if (txn.req && txn.req.state) { tok['state'] = txn.req.state; }
        
        doIssueIDToken(tok);
      }
      
      try {
        // NOTE: To facilitate code reuse, the `issueToken` callback should
        //       interoperate with the `issue` callback implemented by
        //       `oauth2orize.grant.token`.
        
        var arity = issueToken.length;
        if (arity == 6) {
          issueToken(txn.client, txn.user, txn.res, txn.req, txn.locals, issued);
        } else if (arity == 5) {
          issueToken(txn.client, txn.user, txn.res, txn.req, issued);
        } else if (arity == 4) {
          issueToken(txn.client, txn.user, txn.res, issued);
        } else { // arity == 3
          issueToken(txn.client, txn.user, issued);
        }
      } catch (ex) {
        return next(ex);
      }
    }
    
    doIssueToken();
  }

  function errorHandler(err, txn, res, next) {
    var mode = 'fragment'
    , params = {}
    , respond;
    if (txn.req && txn.req.responseMode) {
      mode = txn.req.responseMode;
    }
    respond = modes[mode];

    if (!respond) {
      return next(err);
    }
    if (respond && respond.validate) {
      try {
        respond.validate(txn);
      } catch(ex) {
        return next(err);
      }
    }

    params.error = err.code || 'server_error';
    if (err.message) { params.error_description = err.message; }
    if (err.uri) { params.error_uri = err.uri; }
    if (txn.req && txn.req.state) { params.state = txn.req.state; }
    return respond(txn, res, params);
  }

  /**
   * Return `id_token token` grant module.
   */
  var mod = {};
  mod.name = 'id_token token';
  mod.request = request;
  mod.response = response;
  mod.error = errorHandler;
  return mod;
}


lib/request/extensions.js
/**
 * Module dependencies.
 */
var AuthorizationError = require('@passport-next/oauth2orize').AuthorizationError;

/**
 * Parse request parameters defined by OpenID Connect.
 *
 * This module is a wildcard parser that parses authorization requests for
 * extensions parameters defined by OpenID Connect.
 *
 * Examples:
 *
 *     server.grant(openid.extensions());
 *
 * References:
 *  - [OpenID Connect Basic Client Profile 1.0 - draft 28](http://openid.net/specs/openid-connect-basic-1_0.html)
 *  - [OpenID Connect Implicit Client Profile 1.0 - draft 11](http://openid.net/specs/openid-connect-implicit-1_0.html)
 *  - [OpenID Connect Messages 1.0 - draft 20](http://openid.net/specs/openid-connect-messages-1_0.html)
 *
 * @return {Object} module
 * @api public
 */
module.exports = function() {
  
  function request(req) {
    var q = req.query
      , ext = {};
    
    // TODO: Only parse these if scope includes `openid`

    function parse(param, split) {
      if (!q[param]) { return; }

      if (typeof q[param] !== 'string') {
        throw new AuthorizationError('Failed to parse ' + param + ' as string', 'invalid_request');
      }

      return (split) ? q[param].split(' ') : q[param];
    }

    ext.nonce = parse('nonce');
    ext.display = parse('display') || 'page';
    if (q.prompt) { ext.prompt = parse('prompt', true); }
    if (q.max_age) { ext.maxAge = parseInt(q.max_age); }
    if (q.ui_locales) { ext.uiLocales = parse('ui_locales', true); }
    if (q.claims_locales) { ext.claimsLocales = parse('claims_locales', true); }
    ext.idTokenHint = q.id_token_hint;
    ext.loginHint = q.login_hint;
    if (q.acr_values) { ext.acrValues = parse('acr_values', true); }
    
    if (q.claims) {
      try {
        ext.claims = JSON.parse(q.claims);
      } catch (_) {
        throw new AuthorizationError('Failed to parse claims as JSON', 'invalid_request');
      }
    }
    
    if (q.registration) {
      try {
        ext.registration = JSON.parse(q.registration);
      } catch (_) {
        throw new AuthorizationError('Failed to parse registration as JSON', 'invalid_request');
      }
    }
    
    // NOTE: The below parameters should be implemented in a separate extension, tracking
    //       the following IETF draft:
    //       https://tools.ietf.org/html/draft-ietf-oauth-jwsreq-06
    //       http://openid.net/specs/openid-connect-core-1_0.html#JWTRequests
    // TODO: Add support for "request" parameter
    // TODO: Add support for "request_uri" parameter
    
    if (ext.prompt && ext.prompt.length > 1) {
      if (ext.prompt.indexOf('none') != -1) { throw new AuthorizationError('Prompt includes none with other values', 'invalid_request'); }
    }
    
    return ext;
  }
  
  var mod = {};
  mod.name = '*';
  mod.request = request;
  return mod;
}


lib/grant/codeIdTokenToken.js
/**
 * Module dependencies.
 */
var _ = require('lodash')
  , AuthorizationError = require('../errors/authorizationerror');


/**
 * Handles requests to obtain a response with an access token, authorization
 * code, and ID token.
 *
 * References:
 *  - [OpenID Connect Standard 1.0 - draft 21](http://openid.net/specs/openid-connect-standard-1_0.html)
 *  - [OpenID Connect Messages 1.0 - draft 20](http://openid.net/specs/openid-connect-messages-1_0.html)
 *  - [OAuth 2.0 Multiple Response Type Encoding Practices - draft 08](http://openid.net/specs/oauth-v2-multiple-response-types-1_0.html)
 *
 * @param {Object} options
 * @param {Function} issue
 * @return {Object} module
 * @api public
 */
module.exports = function(options, issueToken, issueCode, issueIDToken) {
  if (typeof options == 'function') {
    issueIDToken = issueCode;
    issueCode = issueToken;
    issueToken = options;
    options = undefined;
  }
  options = options || {};
  
  if (!issueToken) throw new TypeError('oauth2orize-openid.codeIDTokenToken grant requires an issueToken callback');
  if (!issueCode) throw new TypeError('oauth2orize-openid.codeIDTokenToken grant requires an issueCode callback');
  if (!issueIDToken) throw new TypeError('oauth2orize-openid.codeIDTokenToken grant requires an issueIDToken callback');

  var modes = options.modes || {};
  if (!modes.fragment) {
    modes.fragment = require('../response/fragment');
  }

  // For maximum flexibility, multiple scope spearators can optionally be
  // allowed.  This allows the server to accept clients that separate scope
  // with either space or comma (' ', ',').  This violates the specification,
  // but achieves compatibility with existing client libraries that are already
  // deployed.
  var separators = options.scopeSeparator || ' ';
  if (!Array.isArray(separators)) {
    separators = [ separators ];
  }
  
  
  /* Parse requests that request `code id_token token` as `response_type`.
   *
   * @param {http.ServerRequest} req
   * @api public
   */
  function request(req) {
    var clientID = req.query['client_id']
      , redirectURI = req.query['redirect_uri']
      , scope = req.query['scope']
      , state = req.query['state']
      , nonce = req.query['nonce'];
      
    if (!clientID) { throw new AuthorizationError('Missing required parameter: client_id', 'invalid_request'); }
    if (typeof clientID !== 'string') { throw new AuthorizationError('Invalid parameter: client_id must be a string', 'invalid_request'); }
    
    if (!nonce) { throw new AuthorizationError('Missing required parameter: nonce', 'invalid_request'); }
    if (typeof nonce !== 'string') { throw new AuthorizationError('Invalid parameter: nonce must be a string', 'invalid_request'); }

    if (scope) {
      if (typeof scope !== 'string') {
        throw new AuthorizationError('Invalid parameter: scope must be a string', 'invalid_request');
      }
      
      for (var i = 0, len = separators.length; i < len; i++) {
        var separated = scope.split(separators[i]);
        // only separate on the first matching separator.  this allows for a sort
        // of separator "priority" (ie, favor spaces then fallback to commas)
        if (separated.length > 1) {
          scope = separated;
          break;
        }
      }
      
      if (!Array.isArray(scope)) { scope = [ scope ]; }
    }
    
    return {
      clientID: clientID,
      redirectURI: redirectURI,
      scope: scope,
      state: state,
      nonce: nonce
    }
  }
  
  /* Sends responses to transactions that request `code id_token token` as `response_type`.
   *
   * @param {Object} txn
   * @param {http.ServerResponse} res
   * @param {Function} complete
   * @param {Function} next
   * @api public
   */
  function response(txn, res, complete, next) {
    if (next === undefined) {
      next = complete;
      complete = function(cb) { return cb(); };
    }

    var mode = 'fragment'
      , respond;
    if (txn.req && txn.req.responseMode) {
      mode = txn.req.responseMode;
    }
    respond = modes[mode];
    
    if (!respond) {
      // http://lists.openid.net/pipermail/openid-specs-ab/Week-of-Mon-20140317/004680.html
      return next(new AuthorizationError('Unsupported response mode: ' + mode, 'unsupported_response_mode', null, 501));
    }
    if (respond && respond.validate) {
      try {
        respond.validate(txn);
      } catch(ex) {
        return next(ex);
      }
    }

    if (!txn.res.allow) {
      var params = { error: 'access_denied' };
      if (txn.req && txn.req.state) { params.state = txn.req.state; }
      return respond(txn, res, params);
    }
    
    function doIssueIDToken(tok) {
      function issued(err, idToken) {
        if (err) { return next(err); }
        if (!idToken) { return next(new AuthorizationError('Request denied by authorization server', 'access_denied')); }
      
        tok['id_token'] = idToken;

        complete(function(err) {
          if (err) { return next(err); }
          return respond(txn, res, tok);
        });
      }
    
      try {
        // NOTE: To facilitate code reuse, the `issueIDToken` callback should
        //       interoperate with the `issue` callback implemented by
        //       `oauth2orize-openid.grant.idToken`.
        
        var arity = issueIDToken.length;
        if (arity == 7) {
          issueIDToken(txn.client, txn.user, txn.res, txn.req, { accessToken: tok.access_token, authorizationCode: tok.code }, txn.locals, issued);
        } else if (arity == 6) {
          issueIDToken(txn.client, txn.user, txn.res, txn.req, { accessToken: tok.access_token, authorizationCode: tok.code }, issued);
        } else if (arity == 5) {
          issueIDToken(txn.client, txn.user, txn.res, txn.req, issued);
        } else { // arity == 4
          issueIDToken(txn.client, txn.user, txn.req, issued);
        }
      } catch (ex) {
        return next(ex);
      }
    }
    
    function doIssueCode(tok) {
      function issued(err, code) {
        if (err) { return next(err); }
        if (!code) { return next(new AuthorizationError('Request denied by authorization server', 'access_denied')); }
      
        tok['code'] = code;
      
        doIssueIDToken(tok);
      }
      
      try {
        // NOTE: To facilitate code reuse, the `issueCode` callback should
        //       interoperate with the `issue` callback implemented by
        //       `oauth2orize.grant.code`.
        
        var arity = issueCode.length;
        if (arity == 7) {
          var locals = txn.locals || {};
          locals.accessToken = tok.access_token;
          issueCode(txn.client, txn.req.redirectURI, txn.user, txn.res, txn.req, txn.locals, issued);
        } else if (arity == 6) {
          issueCode(txn.client, txn.req.redirectURI, txn.user, txn.res, txn.req, issued);
        } else if (arity == 5) {
          issueCode(txn.client, txn.req.redirectURI, txn.user, txn.res, issued);
        } else { // arity == 4
          issueCode(txn.client, txn.req.redirectURI, txn.user, issued);
        }
      } catch (ex) {
        return next(ex);
      }
    }
    
    function doIssueToken() {
      function issued(err, accessToken, params) {
        if (err) { return next(err); }
        if (!accessToken) { return next(new AuthorizationError('Request denied by authorization server', 'access_denied')); }
      
        var tok = {};
        tok['access_token'] = accessToken;
        if (params) { _.merge(tok, params); }
        tok['token_type'] = tok['token_type'] || 'Bearer';
        if (txn.req && txn.req.state) { tok['state'] = txn.req.state; }
      
        doIssueCode(tok);
      }
      
      try {
        // NOTE: To facilitate code reuse, the `issueToken` callback should
        //       interoperate with the `issue` callback implemented by
        //       `oauth2orize.grant.token`.
        
        var arity = issueToken.length;
        if (arity == 6) {
          issueToken(txn.client, txn.user, txn.res, txn.req, txn.locals, issued);
        } else if (arity == 5) {
          issueToken(txn.client, txn.user, txn.res, txn.req, issued);
        } else if (arity == 4) {
          issueToken(txn.client, txn.user, txn.res, issued);
        } else { // arity == 3
          issueToken(txn.client, txn.user, issued);
        }
      } catch (ex) {
        return next(ex);
      }
    }
    
    doIssueToken();
  }

  function errorHandler(err, txn, res, next) {
    var mode = 'fragment'
    , params = {}
    , respond;
    if (txn.req && txn.req.responseMode) {
      mode = txn.req.responseMode;
    }
    respond = modes[mode];

    if (!respond) {
      return next(err);
    }
    if (respond && respond.validate) {
      try {
        respond.validate(txn);
      } catch(ex) {
        return next(err);
      }
    }

    params.error = err.code || 'server_error';
    if (err.message) { params.error_description = err.message; }
    if (err.uri) { params.error_uri = err.uri; }
    if (txn.req && txn.req.state) { params.state = txn.req.state; }
    return respond(txn, res, params);
  }

  /**
   * Return `code id_token token` grant module.
   */
  var mod = {};
  mod.name = 'code id_token token';
  mod.request = request;
  mod.response = response;
  mod.error = errorHandler;
  return mod;
}
